coalib/core/Core.py

```diff
@@ -327,34 +149,192 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
+                    'please report it to the developers.'.format(', '.join(
+                        repr(dependant) + ' depends on ' + repr(dependency)
+                        for dependency, dependant in self.dependency_tracker)))
+
+            self.event_loop.stop()
+
+    def _finish_task(self, bear, task):
+        """
+        The callback for when a task of a bear completes. It is responsible for
+        checking if the bear completed its execution and the handling of the
+        result generated by the task. It also schedules new tasks if
+        dependencies get resolved.
+
+        :param bear:
+            The bear that the task belongs to.
+        :param task:
+            The task that completed.
+        """
+        try:
+            results = task.result()
+
+            for dependant in self.dependency_tracker.get_dependants(bear):
+                dependant.dependency_results[type(bear)] += results
+        except Exception as ex:
+            # FIXME Try to display only the relevant traceback of the bear if
+            # FIXME   error occurred there, not the complete event-loop
+            # FIXME   traceback.
+            logging.error('An exception was thrown during bear execution.',
+                          exc_info=ex)
+
+            results = None
+
+            # Unschedule/resolve dependent bears, as these can't run any more.
+            dependants = self.dependency_tracker.get_all_dependants(bear)
+            for dependant in dependants:
+                self.dependency_tracker.resolve(dependant)
+            logging.debug('Following dependent bears were unscheduled: ' +
+                          ', '.join(repr(dependant)
+                                    for dependant in dependants))
+        finally:
+            self.running_tasks[bear].remove(task)
+            self._cleanup_bear(bear)
+
+        if results is not None:
+            for result in results:
+                try:
+                    # FIXME Long operations on the result-callback could block
+                    # FIXME   the scheduler significantly. It should be
+                    # FIXME   possible to schedule new Python Threads on the
+                    # FIXME   given event_loop and process the callback there.
+                    self.result_callback(result)
+                except Exception as ex:
+                    # FIXME Try to display only the relevant traceback of the
+                    # FIXME   result handler if error occurred there, not the
+                    # FIXME   complete event-loop traceback.
+                    logging.error(
+                        'An exception was thrown during result-handling.',
+                        exc_info=ex)
+
+
 def run(bears, result_callback):
     """
-    Runs a coala session.
+    It initiates a session with the given parameters and runs it.
 
     :param bears:
-        The bear instances to run.
+            The bear instances to run.
```

Comment:
it's even unnecessary, the additional indentation is not needed^^ (and also inconsistent)